clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java

```diff
@@ -1099,6 +1104,28 @@ private <T extends Serializer<String>> void doTestHeaders(Class<T> serializerCla
         producer.close(Duration.ofMillis(0));
     }
 
+    @SuppressWarnings("unchecked")
+    @Test
+    public void testHeadersFailure() {
+        Properties props = new Properties();
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        props.put(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, ProducerInterceptorForHeaders.class.getName());
+        Serializer<String> keySerializer = mock(Serializer.class);
+        Serializer<String> valueSerializer = mock(Serializer.class);
+
+        KafkaProducer<String, String> producer = new KafkaProducer<>(props, keySerializer, valueSerializer);
+        ProducerRecord<String, String> record = new ProducerRecord<>("topic", "key", "value");
+        Future<RecordMetadata> future = producer.send(record, (recordMetadata, exception) -> { });
+        try {
+            assertInstanceOf(TimeoutException.class, assertThrows(ExecutionException.class, future::get).getCause());
+            //ensure headers are writable if send failure
```

Comment:
This is header of producer record, different from onAcknowledgement's headers.

Producer record headers is writable because client might want to mutate it before retry. This is same behaviour as today, no change introduced from this PR.